File: backend/dotnet/src/Mateu.Core/Naming.cs
using System.Text;

namespace Mateu.Core;

internal static class Naming
{
    /// <summary>"Name" → "name", "FirstName" → "firstName".</summary>
    public static string CamelCase(string s) =>
        string.IsNullOrEmpty(s) ? s : char.ToLowerInvariant(s[0]) + s[1..];

    /// <summary>"firstName" / "FirstName" → "First name".</summary>
    public static string Humanize(string s)
    {
        if (string.IsNullOrEmpty(s)) return s;
        var sb = new StringBuilder();
        for (var i = 0; i < s.Length; i++)
        {
            var c = s[i];
            if (i > 0 && char.IsUpper(c) && !char.IsUpper(s[i - 1])) sb.Append(' ');
            sb.Append(i == 0 ? char.ToUpperInvariant(c) : char.ToLowerInvariant(c));
        }
        return sb.ToString();
    }
}

File: backend/python/mateu_core/naming.py
"""Name conversions, mirroring the C# Naming helper."""

from __future__ import annotations


def camel_case(s: str) -> str:
    """``snake_case`` -> ``camelCase``; ``"name" -> "name"``, ``"first_name" -> "firstName"``."""
    if not s:
        return s
    if "_" not in s:
        return s[0].lower() + s[1:]
    head, *rest = s.split("_")
    return head + "".join(p[:1].upper() + p[1:] for p in rest)


def humanize(s: str) -> str:
    """``first_name`` / ``firstName`` -> ``"First name"`` (matches C# Naming.Humanize)."""
    if not s:
        return s
    out: list[str] = []
    prev_word_char = False  # previous emitted char was a letter/digit (lower)
    for c in s:
        if c == "_":
            if out and out[-1] != " ":
                out.append(" ")
            prev_word_char = False
            continue
        if c.isupper() and prev_word_char:
            out.append(" ")
        out.append(c.lower())
        prev_word_char = c.isalnum()
    text = "".join(out).strip()
    return text[0].upper() + text[1:] if text else text
